### PR TITLE
ddl: Global index more restrictions

### LINKED ISSUE #55424: Can create global index with partition columns
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t(a int, b int) partition by hash(b) partitions 5;
alter table t add unique idx(b) global;
```

### 2. What did you expect to see? (Required)

report error

### 3. What did you see instead (Required)

alter global index success

### 4. What is your TiDB version? (Required)

```
mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                           |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v8.2.0-alpha-869-g1b172b674f
Edition: Community
Git Commit Hash: 1b172b674fa37259fabc4fcbc205274d4091fcce
Git Branch: master
UTC Build Time: 2024-08-15 03:11:41
GoVersion: go1.23.0
Race Enabled: false
Check Table Before Drop: false
Store: unistore |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #55424

Problem Summary:
GLOBAL IndexOption was allowed on non-partitioned tables, non-unique indexes, as well as on indexes including all partitioning columns.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/executor.go [hunk 1]
```diff
@@ -4660,9 +4714,14 @@ func (e *executor) createIndex(ctx sessionctx.Context, ti ast.Ident, keyType ast
 				return dbterror.ErrUniqueKeyNeedAllFieldsInPf.GenWithStackByArgs("UNIQUE INDEX")
 			}
 			// index columns does not contain all partition columns, must set global
-			if indexOption == nil || !indexOption.Global {
+			if !globalIndex {
 				return dbterror.ErrGlobalIndexNotExplicitlySet.GenWithStackByArgs(indexName.O)
 			}
+		} else {
+			// TODO: remove this restriction
+			if globalIndex {
```
Comment (Defined2014) [inline pkg/ddl/executor.go:4722]: better use `else if` here

**Review by Defined2014 — APPROVED** [review]
LGTM

Another question, how about `create table t(a int, unique index idx(a) local)`? Thougth it won't cause any problem, it looks weird.

**Comment by mjonss** [comment]
> LGTM
> 
> Another question, what about `create table t(a int, unique index idx(a) local)`? Thougth it won't cause any problem, it looks weird.

@Defined2014 `local` should just be ignored, I just added it to be compatible with Oracle Syntax, it is just the same as not specifying it. Only `global` is a difference from the default (local).

File: pkg/ddl/executor.go (hunk 1 above)
Comment (mjonss) [reply pkg/ddl/executor.go]: Done.

**Review by tangenta — APPROVED** [review]

**Pull request merged**